### PR TITLE
feat: Dio adaptation for header based auth

### PR DESCRIPTION
**Pull request by iresharma**
## Summary of change
This PR adds header based auth support the dio adaption of this SDK.

## Related issues
None

## Test Plan
Most existing tests fail

## Documentation changes
pending

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `lib/src/version.dart`
- [ ] Changes to the version if needed
   - In `pubspec.yaml`
   - In `lib/src/version.dart`
- [ ] Had installed and ran the pre-commit hook
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.
### Important
This PR is against another PR as it uses a lot of methods and functions implemented in that un-merged PR

## Remaining TODOs for this PR
None

### CODE REVIEW

File: lib/src/dio-interceptor-wrapper.dart [hunk 1]
```diff
@@ -62,6 +64,15 @@ class SuperTokensInterceptorWrapper extends Interceptor {
...
 
+    SuperTokensTokenTransferMethod tokenTransferMethod =
+        SuperTokens.config.tokenTransferMethod!;
+    options.headers["st-auth-mode"] =
+        tokenTransferMethod == SuperTokensTokenTransferMethod.COOKIE
+            ? 'cookie'
+            : 'header';
+
+    options = await _setAuthorizationHeaderIfRequired(options);
+
```
Comment (nkshah2) [inline lib/src/dio-interceptor-wrapper.dart:75]: Need to check for duplicate cookies when combining user defined cookies and those from the cookie manager

File: lib/src/dio-interceptor-wrapper.dart [hunk 2]
```diff
@@ -170,19 +181,66 @@ class SuperTokensInterceptorWrapper extends Interceptor {
...
+        response.headers.map[REFRESH_TOKEN_NAME]?.first.toString();
+
+    if (refreshHeader != null) {
+      Utils.setToken(TokenType.REFRESH, refreshHeader);
+    }
+  }
+
+  Future<RequestOptions> _removeAuthHeaderIfMatchesLocalToken(
+      RequestOptions req) async {
+    if (req.headers.containsKey("Authorization")) {
```
Comment (nkshah2) [inline lib/src/dio-interceptor-wrapper.dart:218]: is containsKey case insensitive?

File: lib/src/front-token.dart [hunk 3]
```diff
@@ -47,13 +47,19 @@ class FrontToken {
             await SuperTokensUtils.getLocalSessionState();
 
         if (localSessionState.status == LocalSessionStateStatus.EXISTS) {
-          _frontTokenMutex.acquire();
+          await _frontTokenMutex.acquire();
         } else {
           finalReturnValue = null;
+          if (_frontTokenMutex.isLocked) {
+            _frontTokenMutex.release();
```
Comment (nkshah2) [inline lib/src/front-token.dart:54]: is release not async?

File: lib/src/front-token.dart [hunk 4]
```diff
@@ -109,12 +115,26 @@ class FrontToken {
       return;
     }
 
-    FrontToken._setFronToken(frontToken);
+    try {
+      await _frontTokenMutex.acquire();
+      FrontToken._setFronToken(frontToken);
```
Comment (nkshah2) [inline lib/src/front-token.dart:120]: does _setFrontToken not need to be awaited?

File: lib/src/supertokens-http-client.dart [hunk 5]
```diff
@@ -281,10 +284,39 @@ class Client extends http.BaseClient {
...
     if (preRequestLocalSessionState.status ==
         LocalSessionStateStatus.NOT_EXISTS) {
-      AntiCSRF.removeToken();
-      FrontToken.removeToken();
+      await AntiCSRF.removeToken();
+      await FrontToken.removeToken();
     }
   }
+
+  static Map<String, dynamic> _generateMapFromCookie(String cookie) {
```
Comment (nkshah2) [inline lib/src/supertokens-http-client.dart:292]: Why use dynamic instead of the cookie type?

Comment (nkshah2) [reply lib/src/supertokens-http-client.dart:292]: Also does the http cookie class not have a method to parse a cookie string?

File: lib/src/supertokens-http-client.dart [hunk 6]
```diff
@@ -281,10 +284,39 @@ class Client extends http.BaseClient {
...
+    String finalCookieHeader = "";
+    existingCookieHeaderMap.keys.forEach((element) {
+      if (newCookiesToAddMap.containsKey(element)) {
+        finalCookieHeader += "$element=${newCookiesToAddMap[element]};";
+      } else {
+        finalCookieHeader += "$element=${existingCookieHeaderMap[element]};";
+      }
+    });
+    return "${finalCookieHeader}HttpOnly;";
+  }
```
Comment (nkshah2) [inline lib/src/supertokens-http-client.dart:319]: Why not do:

1. filter existing cookies in a way that no cookies with the same names exist in newCookiesToAdd
2. Combine newCookiesToAdd and the filtered cookies
3. join with ";"

File: test/dioInterceptor_test.dart [hunk 7]
```diff
@@ -121,10 +120,14 @@ void main() {
     SuperTokens.init(apiDomain: apiBasePath);
     RequestOptions req = SuperTokensTestUtils.getLoginRequestDio();
     Dio dio = setUpDio();
+    print("pre login req");
```
Comment (nkshah2) [inline test/dioInterceptor_test.dart:123]: Remove all print statements

File: test/supertokens_test.dart [hunk 8]
```diff
@@ -42,48 +39,45 @@ void main() {
...
-        fail("API responded with staus ${userInfoResp.statusCode}");
-      }
-    }
-    Uri refreshCustomHeader = Uri.parse("$apiBasePath/refreshHeader");
-    var refreshResponse = await http.get(refreshCustomHeader);
-    if (refreshResponse.statusCode != 200) fail("Refresh Request failed");
-    var respJson = jsonDecode(refreshResponse.body);
-    if (respJson["value"] != "custom-value") fail("Header not sent");
-  });
+  // test("Test custom headers for refreshAPI", () async {
```
Comment (nkshah2) [inline test/supertokens_test.dart:42]: Why was this test commented?

File: testHelpers/server/index.js [hunk 9]
```diff
@@ -418,7 +421,7 @@ app.use("*", async (req, res, next) => {
 app.use(errorHandler());
 
 app.use(async (err, req, res, next) => {
-    console.log(err);
+    // console.log(err);
```
Comment (nkshah2) [inline testHelpers/server/index.js:424]: Remove all console logs from this file. Remove, dont just comment them

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: lib/src/front-token.dart (hunk 3 above)
Comment (iresharma) [reply lib/src/front-token.dart:54]: nope

**Pull request merged**